### PR TITLE
num_samples

### PR DESCRIPTION
**Pull request by zzhhjjj**
generate multiple samples for same input

### CODE REVIEW

File: src/nanotron/generation/decode.py [hunk 1]
```diff
@@ -190,6 +190,16 @@ def decode_text(
 
     p2p = model.p2p
 
+    if generation_config and generation_config.n_samples:
+        assert isinstance(generation_config.n_samples, int) and generation_config.n_samples > 0
```
Comment (NouamaneTazi) [inline src/nanotron/generation/decode.py:194]: this arg is alrdy validated as an int. Remove this line

**Review by NouamaneTazi — APPROVED** [review]
LGTM minus a single comment!

File: src/nanotron/generation/decode.py (hunk 1 above)
Comment (zzhhjjj) [reply src/nanotron/generation/decode.py]: resolved

File: run_generate.py [hunk 2]
```diff
@@ -177,7 +180,7 @@ def main():
             parallel_context=parallel_context,
             max_new_tokens=args.max_new_tokens,
             max_micro_batch_size=2,
-            generation_config=GenerationArgs(sampler="greedy", use_cache=True),
+            generation_config=GenerationArgs(sampler="top_k", top_k=50, use_cache=False, n_samples=2),
```
Comment (NouamaneTazi) [inline run_generate.py:183]: revert this. we should keep greedy as default

File: run_generate.py [hunk 3]
```diff
@@ -164,9 +164,12 @@ def main():
...
         tokenizer.truncation_side = "left"  # TODO @nouamane: do we want this?
         dummy_inputs = [
-            # "Passage: Daniel went back to the garden. Mary travelled to the kitchen. Sandra journeyed to the kitchen. Sandra went to the hallway. John went to the bedroom. Mary went back to the garden. Where is Mary?\nAnswer:",
+            "The future of AI is",
+            "Passage: Daniel went back to the garden. Mary travelled to the kitchen. Sandra journeyed to the kitchen. Sandra went to the hallway. John went to the bedroom. Mary went back to the garden. Where is Mary?\nAnswer:",
             "def fib(n)",
-            # "This film was probably inspired by Godzilla",
+            "This film was probably inspired by Godzilla",
+            "Advancements in technology will lead to",
+            "Tomorrow's world is shaped by",
```
Comment (NouamaneTazi) [inline run_generate.py:172]: maybe add a very long prompt to test as well?
You can maybe get one from here: https://huggingface.co/datasets/HuggingFaceTB/cosmopedia/viewer/web_samples_v2/train?f[text_token_length][min]=2274&f[text_token_length][imax]=2504

File: src/nanotron/generation/decode.py [hunk 4]
```diff
@@ -190,6 +190,16 @@ def decode_text(
...
 
+    # replicate input for n_samples times when using TOP_P or TOP_K samplers, in order to get diverse results
+    if generation_config and generation_config.n_samples:
+        if sampler_type != SamplerType.TOP_P and sampler_type != SamplerType.TOP_K:
+            raise ValueError("Only support n_samples for TOP_P and TOP_K sampler")
+        new_input_iter = []
+        for input in input_iter:
+            for _ in range(generation_config.n_samples):
+                new_input_iter.append(GenerationInput(text=input.text))
+        input_iter = new_input_iter
```
Comment (NouamaneTazi) [inline src/nanotron/generation/decode.py:201]: can you refactor this to be in one line please? (trying to minimize lines of code in src/nanotron 😅)

File: src/nanotron/trainer.py [hunk 5]
```diff
@@ -656,13 +656,10 @@ def init_model(self) -> Union[NanotronModel, DistributedDataParallel]:
...
                 )
             else:
-                log_rank(
-                    f"Setting max_position_embeddings to {self.config.tokens.sequence_length}. Previous value was {self.model_config.max_position_embeddings}.",
-                    logger=logger,
-                    level=logging.INFO,
-                    rank=0,
-                )
-                self.model_config.max_position_embeddings = self.config.tokens.sequence_length
+                # Rather than indirectly setting max_position_embeddings to tokens.sequence_length, we should directly assert their equality.
```
Comment (NouamaneTazi) [inline src/nanotron/trainer.py:659]: this comment won't make sense after this PR is merged. You can delete it

**Review by NouamaneTazi — CHANGES_REQUESTED** [review]

**Review by NouamaneTazi — APPROVED** [review]
LGTM! 🚀

**Pull request merged**